src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,314 +392,191 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// <returns>The message that can be used to write the request operation.</returns>
+        protected abstract ODataBatchOperationRequestMessage CreateOperationRequestMessageImplementation(string method, Uri uri,
+            string contentId, BatchPayloadUriOption payloadUriOption);
 
-            // important to do this first since it will set up the change set boundary.
-            this.SetState(BatchWriterState.ChangeSetStarted);
-            Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
 
-            // reset the size of the current changeset and increase the size of the batch
-            this.ResetChangeSetSize();
-            this.InterceptException(this.IncreaseBatchSize);
+        /// <summary>
+        /// Sets a new writer state; verifies that the transition from the current state into new state is valid.
+        /// </summary>
+        /// <param name="newState">The writer state to transition into.</param>
+        protected void SetState(BatchWriterState newState)
+        {
+            this.InterceptException(() => this.ValidateTransition(
+                newState,
+                () => ValidateTransitionImplementation(newState)));
 
-            // write the boundary string
-            ODataBatchWriterUtils.WriteStartBoundary(this.rawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
-            this.batchStartBoundaryWritten = true;
+            SetStateImplementation(newState);
 
-            // write the change set headers
-            ODataBatchWriterUtils.WriteChangeSetPreamble(this.rawOutputContext.TextWriter, this.changeSetBoundary);
-            this.changesetStartBoundaryWritten = false;
+            this.state = newState;
         }
 
         /// <summary>
-        /// Verifies that calling WriteEndChangeset is valid.
+        /// Additional processing required when setting a new writer state.
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanWriteEndChangeset(bool synchronousCall)
+        /// <param name="newState">The writer state to transition into.</param>
+        protected virtual void SetStateImplementation(BatchWriterState newState)
         {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
         }
 
         /// <summary>
-        /// Ends an active changeset - implementation of the actual functionality.
+        /// Additional validation required when setting a new writer state.
         /// </summary>
-        private void WriteEndChangesetImplementation()
+        /// <param name="newState">The writer state to transition into.</param>
+        protected virtual void ValidateTransitionImplementation(BatchWriterState newState)

```

Comment:
>ValidateTransitionImplementation [](start = 31, length = 32)

As with SetStateImplementation, any format-specific logic in ValidateTransitionImplementation should just be performed in the derived class, before transitioning to the new state.  The derived class can check changeSetBoundary != null before calling SetState(ChangesetStarted) and BatchCompleted, and can verify it's null prior to calling SetState(ChangesetCompleted),  HOWEVER, I think these aren't really format-specific.  I think checking changesetBoundary == null is equivalent to checking IsInsideSubBatch, and the logic should just move to the base classes SetState implementation; i.e., you can't SetState(ChangesetStarted) or SetState(BatchCompleted) if you are inside a subbatch, and can't call SetState(ChangesetCompleted) if you are not inside a subbatch.  So we should be able to remove this protected method and do the state transition validation in SetState.